internal/ingest/priority_queue.go

```diff
@@ -0,0 +1,95 @@
+package ingest
+
+import "sync"
+
+type Queue struct {
+	lk     sync.Mutex
+	states map[providerID]adInfo
+	order  []providerID
+
+	// used if someone is pulling an empty queue
+	waiters    uint
+	notifyChan chan struct{}
+	doneChan   chan struct{}
+}
+
+func NewPriorityQueue() *Queue {
+	return &Queue{
+		states:     make(map[providerID]adInfo),
+		order:      []providerID{},
+		waiters:    0,
+		notifyChan: make(chan struct{}),
+		doneChan:   make(chan struct{}),
+	}
+}
+
+func (q *Queue) Push(p providerID, a adInfo) {
+	q.lk.Lock()
+	defer q.lk.Unlock()
+
+	if _, ok := q.states[p]; !ok {
+		q.order = append(q.order, p)
+	}
+	q.states[p] = a
+
+	if q.waiters > 0 {
+		q.notifyChan <- struct{}{}
```

Comment:
Instead of using `q.waiters` to determine if the channel can be written, what about using select:
```suggestion
		select {
		case q.notifyChan <- struct{}{}:
		default:
		}
```

That should do the same thing and then `q.waiters` is not needed.